vault/eventbus/bus.go

```diff
@@ -50,11 +61,30 @@ func (bus *EventBus) Start() {
 // Send sends an event to the event bus and routes it to all relevant subscribers.
 // This function does *not* wait for all subscribers to acknowledge before returning.
 func (bus *EventBus) Send(ctx context.Context, eventType logical.EventType, data *logical.EventData) error {
+	var nspace string
+	ns, err := namespace.FromContext(ctx)
+	if err == namespace.ErrNoNamespace {
+	} else if err != nil {
+		return err
+	} else {
+		nspace = ns.ID
+	}
 	if !bus.started.Load() {
 		return ErrNotStarted
 	}
+	pluginInfo, ok := ctx.Value(contextEventPluginInfo).(*logical.EventPluginInfo)
```

Comment:
I'd prefer to avoid stuffing additional data into the context if possible. Would it make sense to have a different interface for trusted (internal to the codebase) vs untrusted (plugins) code? Then core's EventBus Send method could include namespace, plugin info, and event data as first class arguments, whereas the plugin interface only includes event data.

Reply:
I went back and forth on this design, whether the core and plugin methods should both be `EventSender`s for simplicity, and using the context to pass the additional plugin information.

I agree that I'd like to not have to use the context if we can avoid it.

I'll refactor this to use a separate method for internal sending and see how we like it.